### DISCUSSION TITLE
How to get test metrics from recorded runs

### DISCUSSION THREAD

**Discussion by parky128** [sorry-cypress/sorry-cypress#396] [Q&A] [answered]
I have just started working with this fantastic set of tools and was able to use the AWS cloud formation template to get up and running pretty easily.

We are using Jenkins and have configured parallel stages to execute, and I can see the completed test runs appearing in our dashboard instance.

After the parallel stages have executed in our Jenkins Pipeline script, I'd like to be able to access the summary information for the test run, e.g. number of failing and passing tests and use that for issuing exit code, etc to fail the build job, comment back in the PR on github etc.

I can see in the Jenkins job output a link is generated for the recorded run.

Is there a way I can query the api using the run identifier to get the metrics I am after for the run?

**Comment by agoldis** [comment, marked as answer]
Thanks for the kind word @parky128. You can send a GraphQL query to achieve the described. Please refer to https://github.com/sorry-cypress/sorry-cypress/blob/0ec6ea72759be59a98a28b1fce75ee46564e187f/packages/api/src/schema/schema.graphql#L14

**Comment by agoldis** [comment]
Alternatively, you can define a "webhook" hook on project-level - it just sends a POST request to an arbitrary URL with the same data.
https://docs.sorry-cypress.dev/integrations/webhooks

**Reply by parky128** [reply]
Thanks @agoldis the webhook approach looks like a good candidate for what I'm wanting to do, I just have no idea how I could use this with my Jenkins Pipeline script - plus I'm a newbie with using webhooks generally but I get the concepts. I have come across their [webhook plugin](https://www.jenkins.io/doc/pipeline/steps/webhook-step/) but it seems to me this will generate a unique hook url that can be registered during the build run, and Im not sure how I could use this with the hook i setup in the project settings on my sorry-cypress dashboard.

I'm probably misunderstanding something fundamental so if you can help give me some pointers here I would be most grateful!

Thanks

**Reply by agoldis** [reply]
For the webhooks - it's just an HTTP request with a payload, nothing more than that. Ideally you'll want to have a static URL that accepts the HTTP request and does whatever you want with the results. One option would be to create an external service that accepts the request and does whatever needs to be done. 

---
If there's no option to create an external service, you can implement a script that sends GraphQL query and fetches the results from the DB. You can alternatively access the DB directly from within the script - it's really up to you.

**Comment by parky128** [comment]
Hi @agoldis 

I'm just looking at this again, and going down the graphQL route

I have a query like so

```
query {
    runs(filters: [{key:"meta.ciBuildId",value:"44"}]){ runId, meta {ciBuildId}, specs { results {stats {suites, passes, failures}}} }
}
```

Returning:

```
{
    "data": {
        "runs": [
            {
                "runId": "2655048fc6da48b8938f1a5ed05c28f6",
                "meta": {
                    "ciBuildId": "44"
                },
                "specs": [
                    {
                        "results": {
                            "stats": {
                                "suites": 1,
                                "passes": 0,
                                "failures": 1
                            }
                        }
                    },
                    {
                        "results": {
                            "stats": {
                                "suites": 1,
                                "passes": 0,
                                "failures": 1
                            }
                        }
                    }
                ]
            }
        ]
    }
}
```

I'd like to get an aggregated count of the stats, e.g overall total of suites, passes, failures.

Can this be done with your graphQL implementation? Is there something built in to graphQL that I could use or does this have to be custom implemented?

Alternatively I can loop over the response object in my own script and generate the overall totals

**Reply by agoldis** [reply]
You'll need to iterate over `specs` list - we also do it for to show the aggregated result for a run on the dashboard.

**Reply by parky128** [reply]
OK thanks, I'll see to that my end then
